src/include/optimizer/logical_operators.h

```diff
@@ -982,5 +983,732 @@ class LogicalExportExternalFile : public OperatorNode<LogicalExportExternalFile>
+ */
+class LogicalCreateNamespace : public OperatorNode<LogicalCreateNamespace> {
+ public:
+  /**
+   * @param namespace_name Name of the namespace to be created
+   * @return
+   */
+  static Operator Make(std::string namespace_name);
+
+  bool operator==(const BaseOperatorNode &r) override;
+  common::hash_t Hash() const override;
+
+  /**
+   * @return the name of the namespace we want to create
+   */
+  const std::string &GetNamespaceName() const { return namespace_name_; }
+
+ private:
+  /**
+   * Name of the new namespace
+   */
+  std::string namespace_name_;
+};
+
+/**
+ * Logical operator for CreateTrigger
+ */
+class LogicalCreateTrigger : public OperatorNode<LogicalCreateTrigger> {
+ public:
+  /**
+   * @param database_oid OID of the database
+   * @param namespace_oid OID of the namespace
+   * @param table_oid OID of the table
+   * @param trigger_name Name of the trigger
+   * @param trigger_funcnames Trigger function names
+   * @param trigger_args Trigger arguments
+   * @param trigger_columns OIDs of trigger columns
+   * @param trigger_when Trigger when clause
+   * @param trigger_type Type of the trigger
+   * @return
+   */
+  static Operator Make(catalog::db_oid_t database_oid, catalog::namespace_oid_t namespace_oid,
+                       catalog::table_oid_t table_oid, std::string trigger_name,
+                       std::vector<std::string> &&trigger_funcnames, std::vector<std::string> &&trigger_args,
+                       std::vector<catalog::col_oid_t> &&trigger_columns,
+                       common::ManagedPointer<parser::AbstractExpression> &&trigger_when, int16_t trigger_type);
+
+  bool operator==(const BaseOperatorNode &r) override;
+  common::hash_t Hash() const override;
+
+  /**
+   * @return OID of the database
+   */
+  const catalog::db_oid_t &GetDatabaseOid() const { return database_oid_; }
+
+  /**
+   * @return OID of the namespace
+   */
+  const catalog::namespace_oid_t &GetNamespaceOid() const { return namespace_oid_; }
+
+  /**
+   * @return OID of the table
+   */
+  const catalog::table_oid_t &GetTableOid() const { return table_oid_; }
+
+  /**
+   * @return trigger name
+   */
+  std::string GetTriggerName() const { return trigger_name_; }
+
+  /**
+   * @return trigger function names
+   */
+  std::vector<std::string> GetTriggerFuncName() const { return trigger_funcnames_; }
+
+  /**
+   * @return trigger args
+   */
+  std::vector<std::string> GetTriggerArgs() const { return trigger_args_; }
+
+  /**
+   * @return trigger columns
+   */
+  std::vector<catalog::col_oid_t> GetTriggerColumns() const { return trigger_columns_; }
+
+  /**
+   * @return trigger when clause
+   */
+  common::ManagedPointer<parser::AbstractExpression> GetTriggerWhen() const {
+    return common::ManagedPointer(trigger_when_);
+  }
+
+  /**
+   * @return trigger type, i.e. information about row, timing, events, access by pg_trigger
+   */
+  int16_t GetTriggerType() const { return trigger_type_; }
```

Comment:
We should make trigger type an enum. We don't need it for TPC-C so it doesn't need to be fixed now.

Can you write an issue about it and we can fix it after we merge this? Mark it as 'beginner'

Reply:
https://github.com/cmu-db/terrier/issues/655